### DISCUSSION TITLE
High-availability DB

### DISCUSSION THREAD

**Discussion by erlend-sh** [trailbaseio/trailbase#9] [Ideas]
You may find [hiqlite](https://github.com/sebadob/hiqlite) interesting.

#### Sidebar: Rauthy as optional auth provider

**Note:** I started this conversation to talk about OAuth providers, but the HA discussion quickly became the center of discussion, so I changed the topic title accordingly.

For people who want to build trailbase apps that can themselves operate as identity providers, Rauthy seems like a natural coupling.

Supabase does this type of integration-partnership as well, for example:
- https://supabase.com/partners/integrations/ory
- https://www.ory.sh/tutorial-url-shortener-supabase-ory-integration-backend/

**Comment by ignatz** [comment]
> For people who want to build trailbase apps that can themselves operate as identity providers, Rauthy seems like a natural coupling.

If I understand correctly, you're not suggesting to "just" use rauthy as an auth provider for a TB app but rather have a TrailBase-Rauthy integration to build auth providers with them? Interesting. I don't see why it wouldn't be possible. I would need to look a bit more to see how TrailBase would even be helpful. I mean rauthy looks pretty sweet and capable already. Curious to hear your thoughts.

> p.s. you may also find [hiqlite](https://github.com/sebadob/hiqlite) interesting

Indeed. I was only aware of https://github.com/rqlite/rqlite and https://github.com/canonical/dqlite, which seem fairly similar. I was planning to look into distribution, replicas and disaster recovery anyway. I might just reach out to the author of rauthy and hiqlite. Curious to hear what they think. They certainly seem to be familiar with the hiqlite alternatives

**Reply by sebadob** [reply]
Hey, this project looks really cool! I was not aware of it yet.

I guess @erlend-sh means to support Auth via Rauthy in case you want to self-host auth as well. So you could start a Rauthy container on the side and use it for logging in to Trailbase. The "natural coupling" makes sense in a way that both projects have simplicity, efficiency and speed in mind.

Making this possible should not  be a big deal. You could either go the simple route and use the [rauthy-client](https://crates.io/crates/rauthy-client) which takes away a lot of (in my eyes) unnecessary abstraction, or use a generic OIDC client crate. The `rauthy-client` should (have not tested it yet though) work with any other OIDC IdP, as long as they support S256 PKCE.

If you need any help with that, feel free to ask.

> Indeed. I was only aware of https://github.com/rqlite/rqlite and https://github.com/canonical/dqlite, which seem fairly similar.

Yes the main idea is the same - make Sqlite highly available and dont have any single point of failure, as this is always a hard requirement for me. I have been experimenting with both of them, and especially `rqlite` is really cool.

The main reason why I decided to create Hiqlite was (as always for me :) ) performance. Both of these projects run Sqlite as a server, but you can't embed rqlite (at least not in Rust) and dqlite is written in C and single threaded. You can create an FFI to a C codebase of course, but then it's still single threaded and you lose Rusts memory safety. Yes, Sqlite is C as well, but the less "unsafe" code, the better.

Hiqlite actually uses quie a few threads under the hood and is highly optimized for speed and throughput. I have not done very detailed benchmarks yet, but you will benefit from more cores when you are below 4 for sure. It get's so far that it is able to get you in a situation where you are physically I/O bound or your network gets saturated with only a Sqlite. In first benchmarks, my Hiqlite setup was able to out-run a Postgres HA cluster (3 nodes, strict replication, to kind of match the same level of guarantees). Only if you run very, very big databases, I would use Postgres, but then on dedicated hosts and so on. Even though I am a big fan of Postgres and used it for many years in both private and enterprise deployments, Sqlite has gotten so powerful in the last years, that it will handle most applications easily with the right configuration.
The way how Hiqlite is set up and configured internally right now, the on-disk Sqlite is actually faster than an in-memory one (in-memory can not use a WAL).

**Reply by ignatz** [reply]
> I guess @erlend-sh means to support Auth via Rauthy in case you want to self-host auth as well. So you could start a Rauthy container on the side and use it for logging in to Trailbase.

@erlend-sh maybe you could clarify? I read "people who want to build trailbase apps that can themselves operate as identity providers", as you were hoping for a deeper integration, e.g. run rauthy and trailbase in the same process with integrated dashboards, ...?

If you only wanted to run Rauthy separately, it should be no different than using Google, Facebook, ... as your auth providers?

> The main reason why I decided to create Hiqlite was (as always for me :) ) performance. Both of these projects run Sqlite as a server, but you can't embed rqlite (at least not in Rust) and dqlite is written in C and single threaded. You can create an FFI to a C codebase of course, but then it's still single threaded and you lose Rusts memory safety. Yes, Sqlite is C as well, but the less "unsafe" code, the better.

That's really helpful, I was wondering what you thought of the alternatives and what made your write your own :+1: . FWIW, I can see why, at least by default, you wouldn't want your HA-db run the same process as your buggy frontend app :). That said, integrating any HA solution into the same binary, would be well in line with TrailBase's mission. I also appreciate the similarity to rusqlite.

Independent of the particular option, I need to think a little more about what's the best compromise especially if folks embed custom Rust/JS logic into TrailBase. Hiqlite definitely seems very promising and flexible enough to draw my own isolation boundaries.

> Only if you run very, very big databases, I would use Postgres, but then on dedicated hosts and so on. 

Just curious, do you have particular limitations in mind? Is this a scalability concern?

Thanks!

**Reply by sebadob** [reply]
> I can see why, at least by default, you wouldn't want your HA-db run the same process as your buggy frontend app

That's not what I meant with "only single threaded". Its more about performance, beacuse even though Sqlite may only allow a single writer at a time, that doesn't mean you get full performance out of it when your app or DB layer is only single threaded. There is so much stuff happning other than just writing to the database, that you basically throw performance out of the window when you think a single thread is enough to saturate Sqlite.

For instance, Hiqlite starts a dedicated thread that only cares about DB modifications, which usually is writing to the DB (and doing `VACCUM`, backups, ...). But everything that thread does in the end is writing to the DB and returning the result to other threads, without interpreting the Result or anything else at all. This way, it can hold the sole lock to the DB and only care about the "writing pipeline" / interface to the DB while all other things like serialization, deserialization, handling results and errors is done on other threads. This approach also makes it possible to get rid of locking datastructures like `Mutex`es and others, which even further reduces the overhead. All internal communication is done via non-locking channels.
Reads are not done on this thread either, they are usually local. So the single writer thread has its own direct connection, even without a connection pool, which is unnecessary in this situation, to reduce the latency even more. For reading however, you will get a small connection pool of (by default) 4 connections. In WAL mode, reads and writes never block each other.

Theoretically, you may be able to get a tiny bit more write throughput out of it, if you would have 2 writer threads, so the other one can always interact with the DB while the first one reads / writes channel messages, but this is a very tiny time window and the 2 threads approach could lead to locks again.

> Just curious, do you have particular limitations in mind? Is this a scalability concern?

This is mostly about scalability. If your app gets so big, that you would run 2 dedicated PG instances with hundreds of gigabytes of memory only for the database, then yes, I would chose Postgres. :)
Apart from that, there are even situations when Sqlite is more stable than Postgres or basically any other network DB, and this is exactly because of that, the network. You can have the most stable DB in the world but it means nothing, if you have network issues.

But, what I am doing in terms of scalability for another application right now, I just start a few different Hiqlite nodes / clients side-by-side. This makes it possible to not only use a new disk for each database, but also a dedicated network. This can be helpful if you have like 4 x 1Gbit connections on your server and 1Gbit is not enough, or if you just want to separate database traffic from everything else of course. This is most probably a huge overkill in this particular example, because a single instance should be able to easily handle all the load, but on the other hand it could scale more at some point while not having too much overhead.

The single writer for Sqlite is usually mentioned as a drawback, but it also has advantages: Writing to the database is much simpler and quicker compared to other ones, because the write lock is set in a single location each time, without any searching through pages and tables only to find the correct location for your table / row lock, and the same when you want to unlock again.

So, the scalability question is not that easily answered. It always depends.

**Reply by ignatz** [reply]
>    I can see why, at least by default, you wouldn't want your HA-db run the same process as your buggy frontend app

Sorry, I might not have expressed myself clearly. My comment wasn't about threading at all. I was thinking that rqlite running HA-sqlite in a different process, i.e. as a client-server architecture, seems like a sensible default. In the single process model (not threads), a buggy application can easily take out some or all nodes in the HA setup. In the client-server model you get a process isolation boundary for free. For TrailBase specifically, you can mix TrailBase and user-code. Even if I tested TrailBase's code most rigorously, there's a chance that user-code will cause e.g. segfault taking down the node.

> For instance, Hiqlite starts a dedicated thread that only cares about DB modifications, which usually is writing to the DB (and doing VACCUM, backups, ...). But everything that thread does in the end is writing to the DB and returning the result to other threads, without interpreting the Result or anything else at all. This way, it can hold the sole lock to the DB and only care about the "writing pipeline" / interface to the DB while all other things like serialization, deserialization, handling results and errors is done on other threads. This approach also makes it possible to get rid of locking datastructures like Mutexes and others, which even further reduces the overhead. All internal communication is done via non-locking channels.
Reads are not done on this thread either, they are usually local. So the single writer thread has its own direct connection, even without a connection pool, which is unnecessary in this situation, to reduce the latency even more. For reading however, you will get a small connection pool of (by default) 4 connections. In WAL mode, reads and writes never block each other.

For reference, TrailBase does exactly that: lock-free worker thread for both writes and reads. I did experiment with keeping reads local but it didn't help performance in my synthetic benchmarks. It certainly made mixed work-loads worse due to lock congestion.
I was initially a bit surprised that local reads didn't help more, since that's what I expected from some other micro-benchmarks. My current hypothesis is that the time TrailBase is spending on sqlite reads is eclipsed by http handling.

> Theoretically, you may be able to get a tiny bit more write throughput out of it, if you would have 2 writer threads, so the other one can always interact with the DB while the first one reads / writes channel messages, but this is a very tiny time window and the 2 threads approach could lead to locks again.

Based on my own experiments, I'm not sure that's practical. Reading / writing channel messages is dirt cheap, it's shared memory after all. It's more likely that one thread will spend almost the entire time polling (or sleeping increasing latency), while the other one is busy.

> The single writer for Sqlite is usually mentioned as a drawback, but it also has advantages: Writing to the database is much simpler and quicker compared to other ones, because the write lock is set in a single location each time, without any searching through pages and tables only to find the correct location for your table / row lock, and the same when you want to unlock again.

I do think it's a limitation for sure but also one that not too many applications will encounter. On a tangent when it comes to horizontal scalability, PG is also not the most scalable option with writes being handled centrally. Minus the finer-grained locking it's not too different from sqlite.

> So, the scalability question is not that easily answered. It always depends.

Very much agreed. Both PG and Sqlite scale pretty well. Once you hit particular bottlnecks you're also more incentivized to look for more specialized solutions.

I do think that many use-cases tend to be more read than write-heavy, so just having read-replicas, as hiqlite provides, goes a long way. Does that make sense?

**Reply by sebadob** [reply]
> a buggy application can easily take out some or all nodes in the HA setup

Yes that's true, but if this is the only application using the database, it doesn't matter, as long as it comes up without issues with the restart. If an app messes up the database, it doesn't matter if its embedded or not, the result will be the same. If you for instance crash a node with embedded Hiqlite, another node will take over the leadership in less than 2 seconds in the default configuration, and this can be reduced down to less than 200ms easily (with reduced max throughput until `openraft-0.10`).

> My current hypothesis is that the time TrailBase is spending on sqlite reads is eclipsed by http handling.

There is probably a pretty high chance that this is the case. When I tested local reads against networked ones, even when on the same host, I was able to reduce the latency by ~10 times, and then even further when I threw away the managed connection pool and created it manually (with Sqlite, there is no need to have them managed imo, because they can't go down because of network issues).

> I do think that many use-cases tend to be more read than write-heavy, so just having read-replicas, as hiqlite provides, goes a long way. Does that make sense?

Absolutely, yes. For a typical web app for instance, it's not uncommon to have 90%+ reads only. When an application has 50 / 50 reads / writes, that's a lot of write throughput already.
But Hiqlite does not only have read replicas in the typical master - slave way. Each node reads locally, yes, but if the leader goes down , any other member can take over that role automatically. I mean, it relies on Raft to do exactly that. And because of the way Raft works, you need to enable synchronous commit in Postgres to achieve the same level of consistency, which makes it even slower for distributed writes.

**Reply by ignatz** [reply]
> Yes that's true, but if this is the only application using the database, it doesn't matter, as long as it comes up without issues with the restart. 

I think you're already one step ahead of me. I was coming from a place of plain sqlite, in which other applications can trivially attach to the same database, in which case this might have been a big "if". However, with hiqlite (or any of the HA solutions) that's no longer possible, since access has to go through a specialized client or the in-process API. And you're saying as long as every participant (e.g. app1, app2, app2-staging, app2-qa) are using the in-process APIs the impact is in the order of hundreds of milliseconds. Got it.

I guess, the integration of other apps that need direct DB access, e.g. because TrailBase rest APIs aren't enough or folks don't want to implement custom endpoints, is also something to keep in mind for the bigger composition picture :thinking: 

> I was able to reduce the latency by ~10 times, and then even further when I threw away the managed connection pool and created it manually

Very similar here. It made some 8x difference in my uBench. Sadly, I didn't manage to make them carry over to TrailBase.

Btw thanks a ton for taking so much time out of your day. Really enjoying the conversation and learning a ton :pray:

**Reply by sebadob** [reply]
> I was coming from a place of plain sqlite, in which other applications can trivially attach to the same database

Ah, yes of course. I mean this is not the best idea because multiple writers can easily block each other, but yeah, makes sense in that case.

> I guess, the integration of other apps that need direct DB access, e.g. because TrailBase rest APIs aren't enough or folks don't want to implement custom endpoints, is also something to keep in mind for the bigger composition picture 🤔

Totally possible with Hiqlite as well. You can embed it, or run it as a server standalone, but in every scenario, you cal always connect via a remote client. It will use the same API the other nodes will use as well.
Hiqlite creates 2 dedicated networks:
- one for raft-internal only traffic and nothing else
- one for a "public" API, and by public I mean it is up to you how public this needs to be.

If you create a cluster with an embedded Hiqlite, nodes will connect to each other via both network layers. The non-internal API is used for all access to the cluster management as well as executing queries. It's the same API no matter how Hiqlite has been started, as long as it's reachable via the network. The client has been designed in a way, that you would use the exact same functions (apart from `query_as_*()` which technically only works for local reads) and it checks internally, which route is the fastest.
This means, if you have multiple apps that need access to the DB, you could run it either as a standalone server, or embed Hiqlite in the "biggest" app or the one that benefits the most from local reads and connect from all others with the remote client.

**Reply by ignatz** [reply]
That makes a lot of sense.  I do think, however, that it won't be quite that easy at least for a while. I'm thinking specifically of users that have other backends or offline batch jobs in other languages, e.g. a php backend that wants connect to the sqlite db, or a python batch job that periodically cleans up some data.

I'm curious, what happens in practice if I `sqlite3` connect to a hiqlite node's DB and mutate it? Reading is probably fine?
